distributed/utils.py

```diff
@@ -1440,3 +1441,91 @@ def __getattr__(name):
         return import_term(use_instead)
     else:
         raise AttributeError(f"module {__name__} has no attribute {name}")
+
+
+class SupportsToDict(Protocol):
+    def to_dict(
+        self, *, exclude: Container[str] | None = None, **kwargs
+    ) -> dict[str, AnyType]:
+        ...
+
+
+@overload
+def recursive_to_dict(
+    obj: SupportsToDict, exclude: Container[str] = None, seen: set[AnyType] = None
+) -> dict[str, AnyType]:
+    ...
+
+
+@overload
+def recursive_to_dict(
+    obj: Sequence, exclude: Container[str] = None, seen: set[AnyType] = None
+) -> Sequence:
+    ...
+
+
+@overload
+def recursive_to_dict(
+    obj: dict, exclude: Container[str] = None, seen: set[AnyType] = None
+) -> dict:
+    ...
+
+
+@overload
+def recursive_to_dict(
+    obj: None, exclude: Container[str] = None, seen: set[AnyType] = None
+) -> None:
+    ...
+
+
+def recursive_to_dict(obj, exclude=None, seen=None):
```

Comment:
I'm generally fond of making things private if they're not used anywhere but for this I'm having a hard time.

Ideally, I want underscored, private functions to be local implementation details, they should not be reused outside of the current scope. As soon as you import or call it out of scope, this is already not true anymore.

in an ideal world I would like to avoid `from distributed.utils import _private_stuff` in favour of `from distributed._private_utils import stuff`

This is also how some IDEs work. If I do the first, my IDE issues a warning and I think that's a fair warning.